src/artemis/external_interaction/tests/test_fgs_callback_collection.py

```diff
@@ -0,0 +1,170 @@
+from unittest.mock import MagicMock
+
+import bluesky.plan_stubs as bps
+import bluesky.preprocessors as bpp
+import pytest
+from bluesky.run_engine import RunEngine
+from ophyd.sim import SynSignal
+
+from artemis.devices.eiger import EigerDetector
+from artemis.devices.fast_grid_scan_composite import FGSComposite
+from artemis.external_interaction.communicator_callbacks import (
+    FGSCallbackCollection,
+    ISPyBDepositionNotMade,
+)
+from artemis.fast_grid_scan_plan import run_gridscan_and_move
+from artemis.parameters import (
+    ISPYB_PLAN_NAME,
+    SIM_BEAMLINE,
+    DetectorParams,
+    FullParameters,
+)
+from artemis.utils import Point3D
+
+
+def test_callback_collection_init(
+    wait_for_result: MagicMock,
+    run_end: MagicMock,
+    run_start: MagicMock,
+):
+    callbacks = FGSCallbackCollection.from_params(FullParameters())
+    assert callbacks.ispyb_handler.params == FullParameters()
+    assert callbacks.zocalo_handler.ispyb == callbacks.ispyb_handler
+
+
+def test_callback_collection_subscription_order_triggers_ispyb_before_zocalo(
+    nexus_writer: MagicMock,
+    mock_ispyb_begin_deposition: MagicMock,
+    mock_ispyb_end_deposition: MagicMock,
+    wait_for_result: MagicMock,
+    run_end: MagicMock,
+    run_start: MagicMock,
+):
+    RE = RunEngine({})
+
+    mock_ispyb_begin_deposition.return_value = ([1, 2], None, 4)
+
+    fgs_undulator_gap = SynSignal(name="fgs_undulator_gap")
+    fgs_synchrotron_machine_status_synchrotron_mode = SynSignal(
+        name="fgs_synchrotron_machine_status_synchrotron_mode"
+    )
+    fgs_slit_gaps_xgap = SynSignal(name="fgs_slit_gaps_xgap")
+    fgs_slit_gaps_ygap = SynSignal(name="fgs_slit_gaps_ygap")
+    detector = SynSignal(name="detector")
+
+    callbacks = FGSCallbackCollection.from_params(FullParameters())
+    callbacklist_right_order = [
+        callbacks.nexus_handler,
+        callbacks.ispyb_handler,
+        callbacks.zocalo_handler,
+    ]
+    assert callbacklist_right_order == list(callbacks)
+
+    @bpp.subs_decorator(list(callbacks))
+    @bpp.run_decorator()
+    def fake_plan():
+        yield from bps.create(ISPYB_PLAN_NAME)
+        yield from bps.read(fgs_undulator_gap)
+        yield from bps.read(fgs_synchrotron_machine_status_synchrotron_mode)
+        yield from bps.read(fgs_slit_gaps_xgap)
+        yield from bps.read(fgs_slit_gaps_ygap)
+        yield from bps.save()
+        yield from bps.read(detector)
```

Comment:
yes, we need to read from something here - otherwise it is the end of the run and the event document is not sent in the format we expect. have added a comment